### PR TITLE
feat(filtering): add SimilarityFilterBlock for near-duplicate removal

### PR DESCRIPTION
**Pull request by UgaTheDev**
## Summary

Adds a standalone `SimilarityFilterBlock` to `core/blocks/filtering/` that removes near-duplicate rows from a DataFrame based on text similarity.

This is a follow-up to #652, where this block was originally part of a larger InstructLab Q&A pipeline PR. Per the [maintainer feedback](https://github.com/Red-Hat-AI-Innovation-Team/sdg_hub/pull/652#issuecomment-4184277191):

> *"the SimilarityFilterBlock is genuinely useful and framework-agnostic. We'd welcome a separate PR for that block on its own (with its tests). It's a clean addition to core/blocks/filtering/ that benefits all users."*

This PR contains **only** the similarity filter block and its tests — no InstructLab pipeline, no qna.yaml references, no flow YAML, no formatter block.

## Design

The block follows the existing `ColumnValueFilterBlock` pattern:

- Inherits from `BaseBlock` (Pydantic)
- Registered via `@BlockRegistry.register("SimilarityFilterBlock", "filtering", ...)`
- Operates on `pd.DataFrame` input/output
- Uses `difflib.SequenceMatcher` for similarity — **zero new dependencies**

### Parameters

| Parameter | Type | Default | Description |
|-----------|------|---------|-------------|
| `input_cols` | `str \| list[str]` | required | Column(s) to compare for similarity (first column is used) |
| `threshold` | `float` | `0.85` | Similarity ratio (0.0–1.0). Rows above this vs any kept row are dropped |
| `group_by` | `str \| None` | `None` | If set, deduplication is scoped within groups of this column |

### Algorithm

1. Extract the comparison column from `input_cols`
2. If `group_by` is set and column exists, group by that column
3. Within each group (or globally), iterate rows and compare against kept rows using `SequenceMatcher.ratio()`
4. Drop rows exceeding the similarity threshold
5. Return filtered DataFrame with reset index

## Files Changed

| File | Change |
|------|--------|
| `src/sdg_hub/core/blocks/filtering/similarity_filter.py` | New block (144 lines) |
| `src/sdg_hub/core/blocks/filtering/__init__.py` | Added export |
| `tests/blocks/filtering/test_similarity_filter.py` | 8 unit tests |

## Usage Example

```python
from sdg_hub.core.blocks.filtering import SimilarityFilterBlock

block = SimilarityFilterBlock(
    block_name="deduplicate_questions",
    input_cols=["question"],
    threshold=0.85,
    group_by="document_id",  # optional
)
filtered_df = block.generate(df)
```

In flow YAML:
```yaml
- block_type: SimilarityFilterBlock
  block_name: deduplicate_questions
  input_cols:
    - question
  threshold: 0.85
  group_by: document_id
```

## Test Plan

- [x] 8 unit tests covering:
  - Unique rows preserved
  - Exact duplicates removed
  - Near-duplicates caught at threshold
  - `group_by` isolates groups correctly
  - `group_by` deduplicates within same group
  - Empty DataFrame returns empty
  - Lower threshold is more aggressive than higher
  - Threshold boundary behavior (0.0)
- [x] No changes to existing code beyond the `__init__.py` export

## Summary by CodeRabbit

* **New Features**
  * Added a similarity-based filtering block to identify and remove duplicate or near-duplicate records.
  * Configurable similarity threshold (0–1) and optional group-by mode to deduplicate within specified groups; preserves first occurrence and returns a reindexed result.

* **Tests**
  * Added comprehensive tests covering unique rows, exact/near duplicates, group-by isolation, empty inputs, threshold boundaries, validation and error cases.

### CODE REVIEW

File: src/sdg_hub/core/blocks/filtering/similarity_filter.py [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+
+        if self.group_by and self.group_by not in samples.columns:
+            logger.warning(
+                "SimilarityFilterBlock '%s': group_by column '%s' not found, "
+                "applying global deduplication",
+                self.block_name,
+                self.group_by,
+            )
+
+        if self.group_by and self.group_by in samples.columns:
```
Comment (shivchander) [inline src/sdg_hub/core/blocks/filtering/similarity_filter.py:139]: **Critical: Silent data loss via `groupby()` dropping NaN keys**

`pd.DataFrame.groupby()` silently drops all rows where the group key is `NaN`/`None` by default. Any row with a missing value in the `group_by` column is permanently removed — not because it was a duplicate, but because pandas excluded it from iteration. The `removed` counter will incorrectly attribute these lost rows to "near-duplicate removal."

This is extremely difficult to debug in practice since the log message actively misleads.

**Fix:** Add `dropna=False`:

```python
for _, group in samples.groupby(self.group_by, dropna=False):
```

Optionally, log a warning when NaN values are detected in the group column.

File: src/sdg_hub/core/blocks/filtering/similarity_filter.py [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+        """Initialize derived attributes after Pydantic validation."""
+        super().model_post_init(__context) if hasattr(
+            super(), "model_post_init"
+        ) else None
+        if self.output_cols is None:
+            self.output_cols = []
+
+    @staticmethod
+    def _similarity(a: str, b: str) -> float:
+        """Compute similarity ratio between two strings."""
```
Comment (shivchander) [inline src/sdg_hub/core/blocks/filtering/similarity_filter.py:82]: **Critical: `_similarity("", "")` returns 0.0 — empty duplicates never filtered**

The guard `if not a or not b: return 0.0` treats two empty strings as having zero similarity. This means unlimited empty-string rows pass through the filter undetected — the most obvious duplication case is never caught.

Consider a dataset where an LLM returned empty responses for 50 rows — all identical (empty), but they all survive.

**Fix:**
```python
if not a and not b:
    return 1.0  # Two empty strings are identical
if not a or not b:
    return 0.0  # One empty, one not
```

File: src/sdg_hub/core/blocks/filtering/similarity_filter.py [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+        Returns
+        -------
+        pd.DataFrame
+            Dataset with near-duplicates removed.
+        """
+        if samples.empty:
+            return samples
+
+        input_cols = cast(list[str], self.input_cols)
+        compare_col = input_cols[0]
```
Comment (shivchander) [inline src/sdg_hub/core/blocks/filtering/similarity_filter.py:127]: **Critical: Missing `input_cols` non-empty validator**

If `input_cols=[]`, this line raises a raw `IndexError: list index out of range` with no actionable message. The sibling `ColumnValueFilterBlock` has an explicit `@field_validator("input_cols", mode="after")` that rejects empty lists.

**Fix:** Add a matching validator:
```python
@field_validator("input_cols", mode="after")
@classmethod
def validate_input_cols_not_empty(cls, v):
    if not v or len(v) == 0:
        raise ValueError(
            "SimilarityFilterBlock requires at least one input column"
        )
    return v
```

File: src/sdg_hub/core/blocks/filtering/similarity_filter.py [hunk 4]
```diff
@@ -0,0 +1,153 @@
+# SPDX-License-Identifier: Apache-2.0
+"""Similarity filter block for removing near-duplicate rows.
+
+This module provides a block that computes pairwise text similarity
```
Comment (shivchander) [inline src/sdg_hub/core/blocks/filtering/similarity_filter.py]: **Improvement: "pairwise" is misleading**

The module docstring says "computes pairwise text similarity" but the algorithm is a greedy sequential scan — each row is compared only against previously *kept* rows, not all-pairs. Reordering the input can change which rows survive, which is not the behavior "pairwise" implies.

Same issue on line 35 in the class docstring.

**Suggestion:** Replace with "compares each row against previously retained rows" or "performs greedy sequential deduplication based on text similarity."

File: src/sdg_hub/core/blocks/filtering/similarity_filter.py [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+within groups and drops near-duplicate entries, keeping only the first
+occurrence when similarity exceeds a configurable threshold.
+"""
+
+# Standard
+from difflib import SequenceMatcher
+from typing import Any, Optional, cast
+
+from pydantic import Field
```
Comment (shivchander) [inline src/sdg_hub/core/blocks/filtering/similarity_filter.py]: **Nit: Import grouping**

`pydantic` is a third-party package but is placed under the `# Standard` section. Move `from pydantic import Field` under `# Third Party` alongside `pandas`.

File: src/sdg_hub/core/blocks/filtering/similarity_filter.py [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+        le=1.0,
+    )
+    group_by: Optional[str] = Field(
+        None,
+        description="Column to group by before comparing similarity.",
+    )
+
+    def model_post_init(self, __context: Any) -> None:
+        """Initialize derived attributes after Pydantic validation."""
+        super().model_post_init(__context) if hasattr(
```
Comment (shivchander) [inline src/sdg_hub/core/blocks/filtering/similarity_filter.py]: **Improvement: Unnecessary `hasattr` guard**

`BaseBlock` inherits from `pydantic.BaseModel`, which always has `model_post_init`. This `hasattr` check can never be `False`, and the ternary-as-statement pattern is confusing.

**Simplify to:**
```python
def model_post_init(self, __context: Any) -> None:
    super().model_post_init(__context)
    if self.output_cols is None:
        self.output_cols = []
```

(I note this pattern is copied from `ColumnValueFilterBlock` — it's a pre-existing issue there too, but no need to perpetuate it.)

File: src/sdg_hub/core/blocks/filtering/similarity_filter.py [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+            for _, group in samples.groupby(self.group_by):
+                groups.append(self._deduplicate_group(group, compare_col))
+            result = (
+                pd.concat(groups, ignore_index=True) if groups else samples.iloc[:0]
+            )
+        else:
+            result = self._deduplicate_group(samples, compare_col)
+
+        removed = original_len - len(result)
+        if removed > 0:
```
Comment (shivchander) [inline src/sdg_hub/core/blocks/filtering/similarity_filter.py:150]: **Improvement: No log when zero duplicates found**

When no duplicates are removed, the block completes silently. If someone targets the wrong column or sets a mismatched threshold, there's no feedback. Consider always logging the outcome:

```python
if removed > 0:
    logger.info(...)
else:
    logger.info(
        "SimilarityFilterBlock '%s': no near-duplicates found "
        "(threshold=%.2f, %d rows examined)",
        self.block_name, self.threshold, original_len,
    )
```

File: tests/blocks/filtering/test_similarity_filter.py [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+class TestSimilarityFilterBlock:
+    """Tests for similarity-based deduplication."""
+
+    def test_keeps_unique_rows(self, make_block):
+        """Completely different rows should all be kept."""
+        block = make_block()
+        df = pd.DataFrame(
+            {"text": ["alpha bravo charlie", "delta echo foxtrot", "golf hotel india"]}
+        )
+        result = block.generate(df)
```
Comment (shivchander) [inline tests/blocks/filtering/test_similarity_filter.py]: **Critical: Tests bypass `BaseBlock.__call__()` validation**

All tests in this file call `block.generate(df)` directly instead of `block(df)`. The `__call__` method performs critical validation (`_validate_dataframe_not_empty`, `_validate_columns`, `_validate_output_columns`) that `generate()` skips entirely.

The sibling `test_columnvaluefilter.py` correctly uses `block(dataset)` throughout.

This masks concrete bugs:
- `test_empty_dataframe` expects empty return, but via `__call__`, `EmptyDatasetError` would be raised
- `test_missing_column_raises_error` expects `KeyError`, but via `__call__`, `MissingColumnError` is raised

**Fix:** Change all `block.generate(df)` calls to `block(df)`, then update `test_empty_dataframe` to assert `EmptyDatasetError` and `test_missing_column_raises_error` to assert `MissingColumnError`.

File: tests/blocks/filtering/test_similarity_filter.py [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+        df = pd.DataFrame(
+            {
+                "text": ["same text here", "same text here"],
+                "doc_id": ["doc_a", "doc_a"],
+            }
+        )
+        result = block.generate(df)
+        assert len(result) == 1
+
+    def test_empty_dataframe(self, make_block):
```
Comment (shivchander) [inline tests/blocks/filtering/test_similarity_filter.py:88]: **Critical: This test verifies wrong behavior**

This test asserts that an empty DataFrame returns empty. But when called via `__call__()` (the production path), `BaseBlock._validate_dataframe_not_empty` raises `EmptyDatasetError` before `generate()` is reached.

The existing `test_columnvaluefilter.py` (line 149) correctly asserts `EmptyDatasetError` for empty datasets.

**Fix:**
```python
def test_empty_dataframe(self, make_block):
    block = make_block()
    df = pd.DataFrame({"text": []})
    with pytest.raises(EmptyDatasetError):
        block(df)
```

File: tests/blocks/filtering/test_similarity_filter.py [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+
+    def test_threshold_boundaries(self, make_block):
+        """Threshold of 0 filters any row with non-zero similarity to kept rows."""
+        df = pd.DataFrame({"text": ["aaa", "aaa!", "bbb"]})
+        block_zero = make_block(threshold=0.0)
+        result = block_zero.generate(df)
+        # "aaa" kept, "aaa!" filtered (similarity > 0), "bbb" kept (similarity = 0)
+        assert len(result) == 2
+
+    def test_missing_column_raises_error(self, make_block):
```
Comment (shivchander) [inline tests/blocks/filtering/test_similarity_filter.py:115]: **Fix: Wrong exception type asserted**

This test expects `KeyError` because it calls `generate()` directly (pandas raises it). Via the production `__call__()` path, `BaseBlock._validate_columns` raises `MissingColumnError` with rich context.

**Fix:**
```python
def test_missing_column_raises_error(self, make_block):
    block = make_block(input_cols=["nonexistent"])
    df = pd.DataFrame({"other_col": ["a", "b"]})
    with pytest.raises(MissingColumnError):
        block(df)
```

File: tests/blocks/filtering/test_similarity_filter.py [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+                threshold=1.5,
+            )
+
+    def test_group_by_missing_column_warns(self, make_block, caplog):
+        """When group_by column is missing, should warn and deduplicate globally."""
+        block = make_block(threshold=0.8, group_by="missing_col")
+        df = pd.DataFrame({"text": ["hello world", "hello world", "different"]})
+        result = block.generate(df)
+        assert len(result) == 2
+        assert "group_by column 'missing_col' not found" in caplog.text
```
Comment (shivchander) [inline tests/blocks/filtering/test_similarity_filter.py:137]: **Suggested additional tests:**

1. **NaN/None in `group_by` column** — validates the `dropna=False` fix
2. **NaN/None in comparison column** — `str(None)` produces `"None"` which silently deduplicates unrelated null rows
3. **Verify first-occurrence retention** — current tests only check row counts, not which row survives
4. **`threshold=1.0` boundary** — `> 1.0` is impossible, so this effectively disables filtering (even exact dupes kept)
5. **Single-row DataFrame** — boundary condition for the dedup loop

**Review by shivchander — CHANGES_REQUESTED** [review]
## PR Review Summary

Thanks for this contribution! The `SimilarityFilterBlock` is a clean, focused addition with zero new dependencies and solid foundational test coverage. There are a few issues to address before merging — details in inline comments below.

### Critical Issues (4)

1. **`groupby()` silently drops NaN group keys** — rows with `NaN` in the `group_by` column are silently discarded by pandas, with the log message incorrectly attributing the loss to deduplication.
2. **Tests call `generate()` directly instead of `__call__()`** — bypasses `BaseBlock` validation (`_validate_dataframe_not_empty`, `_validate_columns`), masking incorrect exception types and behavior.
3. **Missing `input_cols` non-empty validator** — empty `input_cols` causes a raw `IndexError` instead of a clear validation error. The sibling `ColumnValueFilterBlock` has this validator.
4. **`_similarity("", "")` returns 0.0** — two empty strings are treated as having zero similarity, so unlimited empty-string duplicate rows pass through unfiltered.

### Important Issues (4)

5. **"pairwise" in docstrings is misleading** — the algorithm is greedy sequential, not all-pairs.
6. **Import grouping** — `pydantic` is a third-party package but sits under `# Standard`.
7. **No logging when zero duplicates found** — silent no-op is indistinguishable from misconfiguration.
8. **`model_post_init` has unnecessary `hasattr` guard** — `BaseBlock` always has `model_post_init`.

### Test Gaps

- No test for `NaN`/`None` values in comparison or `group_by` columns
- No test verifying *which* row is kept (first occurrence) — tests only check counts
- No test for `threshold=1.0` boundary (effectively disables filtering since `> 1.0` is impossible)
- No test for single-row DataFrame

### Strengths

- Clean single-responsibility design, zero new dependencies
- Good Pydantic `Field(ge=0.0, le=1.0)` constraint on threshold
- Useful O(n*k) complexity note with MinHash suggestion
- Solid test fixture pattern and good coverage of core scenarios

**Comment by shivchander** [comment]
Hi @UgaTheDev — thanks for this contribution! The SimilarityFilterBlock is a great idea and we'd love to have it in the project.

Since there's been no activity on the review feedback for a while, we're going to pick this up internally and implement it ourselves, incorporating the suggestions from the review. Your original PR and design will be credited in the follow-up.

Thanks again for the contribution!
